### PR TITLE
FXIOS-1739: start at home

### PR DESCRIPTION
**Pull request by nishant2718**
# Overview

This PR addresses this [JIRA ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-1739).

### CODE REVIEW

**Comment by nishant2718** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

    - The goal of this metric is to understand if users are in general opening up to Firefox Home or a custom set home page. 

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
    - These events will help us understand the usage of Starting at Home, so we can confirm its value for users.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

   None

4) Can current instrumentation answer these questions?

   No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   
Category 2

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Setting Home to a custom URL or Firefox Home is determined.</td>
    <td>Category 2</td>
    <td>https://mozilla-hub.atlassian.net/browse/FXIOS-1739</td>
  </tr>
  <tr>

  <tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
Available in metrics.yaml and https://dictionary.telemetry.mozilla.org/apps/firefox_ios

7) How long will this data be collected?  Choose one of the following:

    Until 2022-01-01 (~6 months) so we can gather enough data.

8) What populations will you measure?
All

9) If this data collection is default on, what is the opt-out mechanism for users?
Initially it'll be hidden behind a feature flag, and not on by default.

10) Please provide a general description of how you will analyze this data.
Using Glean / Amplitude

11) Where do you intend to share the results of your analysis?
Only on Glean, Amplitude and with mobile teams.

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
None

File: Client/Telemetry/TelemetryWrapper.swift [hunk 1]
```diff
@@ -199,6 +199,18 @@ class TelemetryWrapper {
...
+        switch homePageSetting {
+        case .topSites:
+            let firefoxHome = "Home"
+            GleanMetrics.Preferences.homePageSetting.set(firefoxHome)
+        case .homePage:
+            let customUrl = "other"
+            GleanMetrics.Preferences.homePageSetting.set(customUrl)
+        default:
+            GleanMetrics.Preferences.homePageSetting.set(homePageSetting.rawValue)
+        }
```
Comment (nishant2718) [inline Client/Telemetry/TelemetryWrapper.swift:213]: Added tracking for Home Page here, where it looks like NewTab is also already tracked.

@dnarcese is newTab tracking correct? I think it might be better to track via `NewTabAccessors.getNewTabPage(prefs)` instead of `NewTabAccessors.Default.rawValue`

Comment (dnarcese) [reply Client/Telemetry/TelemetryWrapper.swift:213]: In the end, its the same thing, right?

File: Client/Frontend/Browser/Tab.swift [hunk 2]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
         }
         return self.url
     }
+    
+    var isFxHomeTab: Bool {
+        if let numberOfUrls = self.sessionData?.urls.count,
+           let offset = self.sessionData?.currentPage,
+           let url = self.sessionData?.urls[numberOfUrls - 1 + offset],
+           url.absoluteString.hasPrefix("internal://") {
+            return true
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:155]: nit: whitespace

File: Client/Frontend/Browser/Tab.swift [hunk 3]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
         }
         return self.url
     }
+    
+    var isFxHomeTab: Bool {
+        if let numberOfUrls = self.sessionData?.urls.count,
+           let offset = self.sessionData?.currentPage,
+           let url = self.sessionData?.urls[numberOfUrls - 1 + offset],
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:153]: could we do a small comment why we're doing `numberOfUrls - 1 + offset`?

File: Client/Frontend/Browser/Tab.swift [hunk 4]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
         }
         return self.url
     }
+    
+    var isFxHomeTab: Bool {
+        if let numberOfUrls = self.sessionData?.urls.count,
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:151]: This is always going to be false for tabs that have been loaded and not restored. This might cause bugs when you haven't used the app but it hasn't been evicted from memory yet. Should we handle that case?

File: Client/Frontend/Browser/Tab.swift [hunk 5]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
...
+           let offset = self.sessionData?.currentPage,
+           let url = self.sessionData?.urls[numberOfUrls - 1 + offset],
+           url.absoluteString.hasPrefix("internal://") {
+            return true
+        }
+        
+        return false
+    }
+    
+    var isCustomHomeSet: Bool {
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:161]: Should this logic live on `tab`? 🤔

File: Client/Frontend/Browser/Tab.swift [hunk 6]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
...
+        }
+        
+        return true
+    }
+    
+    var isCustomHomeTab: Bool {
+        guard let profile = self.browserViewController?.profile else { return false }
+        
+        if let customHomeUrl = HomeButtonHomePageAccessors.getHomePage(profile.prefs),
+           let numberOfUrls = self.sessionData?.urls.count,
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:176]: It might make more sense to have one variable `isAtHome` that does the appropriate check, versus having both and doing the check twice based on the users pref.

File: Client/Frontend/Browser/Tab.swift [hunk 7]
```diff
@@ -146,6 +146,45 @@ class Tab: NSObject {
...
+           let offset = self.sessionData?.currentPage,
+           let url = self.sessionData?.urls[numberOfUrls - 1 + offset],
+           let baseDomain = url.baseDomain,
+           let customHomeBaseDomain = customHomeUrl.baseDomain,
+           baseDomain.hasPrefix(customHomeBaseDomain) {
+            return true
```
Comment (boek) [inline Client/Frontend/Browser/Tab.swift:182]: nit: whitespace

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 8]
```diff
@@ -8,10 +8,11 @@ import Shared
 import XCGLogger
 
 private let log = Logger.browserLogger
-class TabManagerStore {
+class TabManagerStore: FeatureFlagsProtocol {
     fileprivate var lockedForReading = false
     fileprivate let imageStore: DiskImageStore?
     fileprivate var fileManager = FileManager.default
+    fileprivate let profile: Profile
```
Comment (boek) [inline Client/Frontend/Browser/TabManagerStore.swift:15]: I'm not against adding `Profile` as a dependency here, but I would like to make sure we're only doing it if absolutely necessary.

File: Client/Telemetry/TelemetryWrapper.swift [hunk 9]
```diff
@@ -199,6 +199,18 @@ class TelemetryWrapper {
         } else {
             GleanMetrics.Preferences.newTabExperience.set(NewTabAccessors.Default.rawValue)
         }
+        // Record `Home` setting, where Firefox Home is "Home", a custom URL is "other" and blank is "Blank".
+        let homePageSetting = NewTabAccessors.getHomePage(prefs)
+        switch homePageSetting {
+        case .topSites:
+            let firefoxHome = "Home"
+            GleanMetrics.Preferences.homePageSetting.set(firefoxHome)
```
Comment (boek) [inline Client/Telemetry/TelemetryWrapper.swift:207]: If we're only using the string here, I think it is okay just to put it in `.set()`

**Review by boek — CHANGES_REQUESTED** [review]

**Comment by dnarcese** [comment]
R+
# Data Review Form

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, telemetry is documented  at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, Firefox iOS provides data controls in the app settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

This should be re-reviewed w/ Ashley Thomas every 6mo

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Type 2

5) Is the data collection request for default-on or default-off?

Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

File: Client/Frontend/Browser/Tab.swift (hunk 4 above)
Comment (nishant2718) [reply Client/Frontend/Browser/Tab.swift:151]: I think you might have to ELI5 🙊

File: Client/Telemetry/TelemetryWrapper.swift (hunk 1 above)
Comment (nishant2718) [reply Client/Telemetry/TelemetryWrapper.swift:213]: While debugging, I noticed that it always returns `"HomePage"` regardless of whether New Tab is set to Firefox Home or Blank. It even returns "Home Page" when a custom Home doesn't exist in that list. 

By the way, I understood this to measure users who have set New Tab in Settings. If this isn't measuring that, you can ignore these comments 🙈

File: Client/Telemetry/TelemetryWrapper.swift (hunk 9 above)
Comment (nishant2718) [reply Client/Telemetry/TelemetryWrapper.swift:207]: (Sorry if I misunderstood!) I had to adjust it slightly here b/c Firefox Home returns "TopSites", a custom set home returns "HomePage" and blank returns "Blank." But what we want is: Firefox Home --> "Home", Custom Set Url --> "other" and Blank --> "Blank"

File: Client/Frontend/Browser/TabManagerStore.swift (hunk 8 above)
Comment (nishant2718) [reply Client/Frontend/Browser/TabManagerStore.swift:15]: After looking through it a bit, I think we'll need to pass `profile` in for logic in `shouldOpenHome`. We can also pass just `prefs` from profile as a dependency. What do you think?

File: Client/Frontend/Browser/Tab.swift (hunk 4 above)
Comment (boek) [reply Client/Frontend/Browser/Tab.swift:151]: @nishant2718 Sorry 😅. When a tab loads the the first time sessionData gets nil'd out https://github.com/mozilla-mobile/firefox-ios/blob/318b942069c61006a621878687a927facfb3560d/Client/Frontend/Browser/Tab.swift#L328.

File: Client/Frontend/Browser/TabManagerStore.swift (hunk 8 above)
Comment (boek) [reply Client/Frontend/Browser/TabManagerStore.swift:15]: @nishant2718 I'm good with that, lets just pass the only dependency that is needed 👍

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 10]
```diff
@@ -128,19 +141,47 @@ class TabManagerStore {
...
         for savedTab in savedTabs {
             // Provide an empty request to prevent a new tab from loading the home screen
             var tab = tabManager.addTab(flushToDisk: false, zombie: true, isPrivate: savedTab.isPrivate)
             tab = savedTab.configureSavedTabUsing(tab, imageStore: imageStore)
             if savedTab.isSelected {
                 tabToSelect = tab
             }
+            
+            if fxHomeTab == nil {
+                tab.isFxHomeTab ? fxHomeTab = tab : nil
```
Comment (boek) [inline Client/Frontend/Browser/TabManagerStore.swift:156]: nit: Use the ternary to return the value to be assigned, not assign inside the ternary
`fxHomeTab = tab.isFxHomeTab ? tab : nil`

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 11]
```diff
@@ -128,19 +141,47 @@ class TabManagerStore {
...
+            }
+            if customHomeTab == nil {
+                tab.isCustomHomeTab ? customHomeTab = tab : nil
```
Comment (boek) [inline Client/Frontend/Browser/TabManagerStore.swift:159]: Same as above :)

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 12]
```diff
@@ -128,19 +141,47 @@ class TabManagerStore {
...
         }
-
+        
+        if shouldOpenHome {
+            let page = NewTabAccessors.getHomePage(prefs)
+            let customUrl = HomeButtonHomePageAccessors.getHomePage(prefs)
+            let homeUrl = URL(string: "internal://local/about/home")
+            
+            if page == .homePage, let customUrl = customUrl {
+                return customHomeTab != nil ? customHomeTab : tabManager.addTab(URLRequest(url: customUrl))
```
Comment (boek) [inline Client/Frontend/Browser/TabManagerStore.swift:173]: I think this could be `customHomeTab ?? tabManager.addTab(URLRequest(url: customURL))`

File: Client/Frontend/Browser/TabManagerStore.swift [hunk 13]
```diff
@@ -128,19 +141,47 @@ class TabManagerStore {
...
+            } else if page == .topSites, let homeUrl = homeUrl {
+                let home = fxHomeTab != nil ? fxHomeTab : tabManager.addTab()
```
Comment (boek) [inline Client/Frontend/Browser/TabManagerStore.swift:175]: same as above :)

**Review by boek — APPROVED** [review]
looks good!!
